### PR TITLE
Add 2FA status and backup codes to Accounts admin

### LINKED ISSUE #519: Add 2FA status indicator and backup code generation to Accounts admin view
## Context

2FA is **required on every `@nobodies.team` account** — it's enforced at the Workspace org policy level, not optional. So the 2FA state we care about is whether each individual account has actually completed enrollment. An account showing as "not enrolled" is a broken account that can't sign in, not a policy-compliance question.

The `/Google/Accounts` admin view (`src/Humans.Web/Views/Google/Accounts.cshtml`) currently surfaces Suspend / Reactivate / Link / Reset Password (#518) but gives Admin zero visibility into 2SV enrollment state. When someone loses their 2FA device there's also no in-app path to issue backup codes — Admin has to drop into the Workspace console.

## Motivation

- **Spot broken accounts before the human reports it.** Surface any `@nobodies.team` account that hasn't completed 2SV enrollment — these people can't sign in.
- **Recover locked-out humans without the Workspace console.** Generate backup verification codes directly from the Accounts page and hand them off.
- Keep Admin work in one place instead of bouncing between Humans and Workspace admin.

### Read: 2SV enrollment per account
- Extend `IGoogleWorkspaceUserService` with a call that reads `isEnrolledIn2Sv` from the Google Admin SDK Directory API (`users.get`). We can ignore `isEnforcedIn2Sv` since enforcement is org-wide.
- Add to `WorkspaceEmailAccountViewModel`.
- Per-row status: a green "2FA ready" chip for enrolled, a red "2FA not set up — account unusable" chip for not enrolled.
- Top-of-page summary counter: "X accounts missing 2FA setup" with a quick filter to show just those.

### Write: generate backup verification codes
- **Generate backup codes** button per row, Admin-only, CSRF-protected.
- Calls Google Admin SDK `verificationCodes.generate` then `verificationCodes.list` to return the freshly issued codes. Google reissues the full set on generate — the confirmation dialog must state that any previously-issued codes will be invalidated.
- Displays codes once in a modal with copy-to-clipboard and a "I've delivered these securely" acknowledgement. Modal closes and page reloads without codes visible again.
- Button is disabled (with tooltip) if the account is not enrolled — backup codes require enrollment to be meaningful.
- Separate **Invalidate backup codes** button (calls `verificationCodes.invalidate`) for when codes may have leaked.

### Authorization + audit
- Admin-only (matches the rest of `/Google/Accounts/*`).
- Generate and invalidate actions both written to the audit log via `IAuditLogService` — admin user id, target email, action. Backup codes are a sensitive surface and need a paper trail.
- Flash messages via existing `SetSuccess` / `SetError`.

### UI / localization
- Status chip label, button labels, confirmation dialog copy, backup-code modal copy localized across `SharedResource.*.resx`.
- Warning copy: "These codes replace any previously issued codes. Deliver them to {email} securely and do not store them."

## Open design questions (minor)

- Show codes in the UI vs. email directly to the recovery email? Recommendation: show-once-in-UI, because the typical flow is Admin helping a human in real time (phone lost, Signal/WhatsApp call) and the human may not have access to their recovery inbox at that moment. Flag in design if that's wrong.

## Acceptance Criteria

- [ ] `WorkspaceEmailAccountViewModel` exposes `IsEnrolledIn2Sv`.
- [ ] Accounts view shows a 2FA enrollment chip per row with clear visual distinction between "ready" (enrolled) and "not set up" (unenrolled, needs admin attention).
- [ ] Summary counter + quick filter for unenrolled accounts.
- [ ] "Generate backup codes" button visible per row, Admin-only, disabled when not enrolled with explanatory tooltip.
- [ ] Clicking confirms with a note that previously-issued codes will be invalidated, then calls `POST /Google/Accounts/GenerateBackupCodes` backed by `IGoogleWorkspaceUserService.GenerateBackupCodesAsync`.
- [ ] Generated codes appear once in a modal with copy-to-clipboard, hidden after dismissal.
- [ ] "Invalidate backup codes" button available; confirmation notes immediate invalidation.
- [ ] Generate + invalidate both written to audit log.
- [ ] All new UI text localized in `SharedResource.*.resx`.

**Key files:**
- `src/Humans.Web/Views/Google/Accounts.cshtml`
- `src/Humans.Web/Controllers/GoogleController.cs`
- `src/Humans.Application/Interfaces/IGoogleWorkspaceUserService.cs`
- `src/Humans.Infrastructure/Services/GoogleWorkspaceUserService.cs` (+ `StubGoogleWorkspaceUserService.cs`)
- `src/Humans.Web/Models/WorkspaceEmailAccountViewModel.cs` (or equivalent)
- `src/Humans.Web/Resources/SharedResource.*.resx`

## Related

- #518 — Password reset button on the same page.
- #111 — Original @nobodies.team account management.

### PR DESCRIPTION
**Pull request by peterdrier**
Closes nobodies-collective/Humans#519

## Summary

`/Google/Accounts` admin page gains the recovery surface that lets admins unblock locked-out @nobodies.team humans without leaving Humans.

**Visibility on every row:**
- 2-Step Verification chip — green "2FA ready" or red "2FA not set up" — backed by Directory API `isEnrolledIn2Sv`
- Recovery email column (yellow "None" badge when unset) — backed by Directory API `recoveryEmail`
- Top-of-page "Missing 2FA" counter + alert banner + filter toggle for the active-unenrolled set (suspended accounts excluded — they can't sign in either way)

**Two recovery actions:**
- **Reset Password** — issues a fresh temp password, surfaces it once in a modal with copy-to-clipboard formatted `pw: <password>` and an "I've delivered these securely" gate
- **Reset + 2FA** — composes password-reset + a single backup verification code in one click. Confirmed working for unenrolled accounts in production. Clipboard format:
  ```
  pw: <temp password>
  2fa: <single backup code>
  ```

The old standalone "Generate / Invalidate backup codes" buttons (and the Backup Codes column) are gone — operationally we only ever needed one code at a time, and the admin always wants the password alongside it.

## Notes

- **Localization:** Per the repo's coding rule (`admin pages do not require localization`), new UI copy is in English only — consistent with the rest of `/Google/Accounts`.
- The stub directory client returns deterministic placeholder credentials so the modal flow is testable locally without Google credentials.
- No new DB migrations. `WorkspaceAccountBackupCodesInvalidated` is left in the audit enum (marked unused) so existing audit-log values don't shift.
- New scope: `admin.directory.user.security` — required by `verificationCodes.generate/list`.

## Test plan

- [ ] `/Google/Accounts` as Admin: 2FA chip visible per row, Recovery Email column populated (or "None" badge), summary counter matches active+unenrolled count, filter toggle narrows the list
- [ ] **Reset Password** on an active account: confirm dialog → modal opens with `pw: <generated>`, copy-to-clipboard works, Close gated on the acknowledgement checkbox, password clears from DOM after closing, single audit entry recorded
- [ ] **Reset + 2FA** on an enrolled account: confirm dialog → modal opens with both `pw:` and `2fa:` lines, copy-to-clipboard puts both into the clipboard, two audit entries recorded
- [ ] **Reset + 2FA** on an unenrolled account: same flow as above, codes are usable for first sign-in
- [ ] Reset / Reset + 2FA buttons hidden for suspended accounts
- [ ] Non-Admin cannot reach either POST endpoint (policy-gated)

### CODE REVIEW

File: src/Humans.Application/Services/GoogleIntegration/GoogleAdminService.cs [hunk 1]
```diff
@@ -250,6 +256,63 @@ await _auditLogService.LogAsync(
...
+        string email, Guid actorUserId,
+        CancellationToken ct = default)
+    {
+        try
+        {
+            var codes = await _workspaceUserService.GenerateBackupCodesAsync(email, ct);
+
+            await _auditLogService.LogAsync(
+                AuditAction.WorkspaceAccountBackupCodesGenerated,
+                "WorkspaceAccount", Guid.Empty,
```
[the remark this answers is not in the document]
Comment (peterdrier) [reply src/Humans.Application/Services/GoogleIntegration/GoogleAdminService.cs]: Fixed in da4ad415. The audit-log call is now wrapped in its own try/catch with a `LogCritical` on failure — codes always flow back to the admin even if audit persistence throws. Added a regression test (`GenerateBackupCodesAsync_ReturnsCodesEvenWhenAuditWriteFails`) that fails the audit-log substitute and asserts `Success: true` with codes intact.

File: src/Humans.Application/Services/GoogleIntegration/GoogleAdminService.cs [hunk 2]
```diff
@@ -315,6 +321,61 @@ await _auditLogService.LogAsync(
...
+                $"Generated {codes.Count} backup code(s) for @{NobodiesTeamDomain} account: {email}",
+                actorUserId);
+
+            return new WorkspaceBackupCodesResult(
+                Success: true,
+                Email: email,
+                Codes: codes,
+                Message: $"Generated {codes.Count} backup code(s) for {email}. Deliver them securely — they cannot be retrieved again.");
```
[the remark this answers is not in the document]
Comment (peterdrier) [reply src/Humans.Application/Services/GoogleIntegration/GoogleAdminService.cs]: Fixed in da4ad415 — guards `codes.Count == 0` before the audit write and returns `Success: false` with an explanatory message, keeping the audit trail honest. Covered by `GenerateBackupCodesAsync_ReturnsFailureAndDoesNotAuditOnEmptyList`.

File: tests/Humans.Application.Tests/Services/GoogleAdminServiceTests.cs [hunk 3]
```diff
@@ -70,9 +70,9 @@ public async Task GetWorkspaceAccountListAsync_ReturnsAccountsWithMatchedUsers()
         _workspaceUserService.ListAccountsAsync(Arg.Any<CancellationToken>())
             .Returns([
                 new WorkspaceUserAccount("alice@nobodies.team", "Alice", "Smith", false,
-                    DateTime.UtcNow, DateTime.UtcNow),
+                    DateTime.UtcNow, DateTime.UtcNow, IsEnrolledIn2Sv: true),
                 new WorkspaceUserAccount("bob@nobodies.team", "Bob", "Jones", true,
-                    DateTime.UtcNow, null),
+                    DateTime.UtcNow, null, IsEnrolledIn2Sv: false),
```
[the remark this answers is not in the document]
Comment (peterdrier) [reply tests/Humans.Application.Tests/Services/GoogleAdminServiceTests.cs]: Done in da4ad415. Added `MissingTwoFactorCount.Should().Be(0)` to the existing test (Bob is suspended → excluded), plus `GetWorkspaceAccountListAsync_CountsActiveUnenrolledTowardMissingTwoFactor` which adds an active+unenrolled Carol and asserts the counter increments to 1. Also added 4 tests for `GenerateBackupCodesAsync` (success, empty list, Workspace failure, audit-write failure) and 2 for `InvalidateBackupCodesAsync` (success, failure).

**Comment by peterdrier** [comment]
Addressed in 6b885ad2.

Both `feature-spec-on-new-feature` and `post-fix-doc-check` warnings landed on the same gap. Fixed in `docs/features/32-workspace-account-provisioning.md` rather than `07-google-integration.md` because file 32 is the topical owner of @nobodies.team account management (its `freshness:triggers` already cover `GoogleWorkspaceUserService.cs` and `Infrastructure/Services/GoogleWorkspace/**`); file 07 is scoped to Drive/Groups sync.

Added:
- **US-32.4** — 2FA enrollment surfacing (chip, summary counter, filter)
- **US-32.5** — Recovery-email surfacing
- **US-32.6** — Backup-codes generate (with the unenrolled-account behavior + graceful-failure path)
- **US-32.7** — Backup-codes invalidate (gated to enrolled accounts)
- New "Account-Management Invariants" section pinning the audit-vs-delivery rationale (empty-list guard, audit-failure preservation) so the ordering isn't only in code comments
- Routes table updated with the three new GET/POST endpoints
- Authorization table updated
- `freshness:flag-on-change` extended to mention the new surface
- IGoogleWorkspaceUserService listing refreshed (was missing Reactivate/ResetPassword/backup-codes methods)
- Drive-by: removed the stale `/Admin/Email` reference (replaced by `/Google/Accounts` in a prior PR)

**Comment by peterdrier** [comment]
## PR Review — 2026-05-04

**PR**: #254 — Add 2FA status and backup codes to Accounts admin
**Branch**: `sprint/2026-04-21/issue-519` → `main`
**Scope**: 17 files, 5 commits, +839 / -38

The latest commit `ea46453c` is a significant redesign (consolidating to **Reset Password** / **Reset + 2FA**) that landed after all prior reviews completed clean on `6b885ad2`. Review focuses on that delta plus verifying earlier-flagged fixes still hold.

---

### Prior Review Status

All three previously-flagged inline items were re-verified against HEAD post-redesign:

- **[Codex P1] Preserve generated codes when audit write fails** — FIXED ✓ (`GoogleAdminService.cs:362-375`, pinned by `GenerateBackupCodesAsync_ReturnsCodesEvenWhenAuditWriteFails`)
- **[Claude] Audit/UI mismatch on empty codes list** — FIXED ✓ (`GoogleAdminService.cs:343-355`, pinned by `GenerateBackupCodesAsync_ReturnsFailureAndDoesNotAuditOnEmptyList`)
- **[Claude] Test coverage for new methods + `MissingTwoFactorCount` assertion** — FIXED ✓ and EXTENDED for the redesign (`GoogleAdminServiceTests.cs:102, 379-543`)
- **[Claude] Feature-spec / freshness-trigger docs gap** — FIXED ✓ (`docs/features/32-workspace-account-provisioning.md` US-32.4 through US-32.7)

---

### Spec Compliance — Issue #519

**Intent**: (1) surface 2SV enrollment per `@nobodies.team` account so admins spot broken accounts; (2) recover locked-out humans without bouncing to the Workspace console; (3) one audit-tracked Admin-only place to do both.

**Verdict: FULLY ADDRESSED via defensible reframing.**

The pivot from "Generate / Invalidate backup codes" buttons to "Reset Password / Reset + 2FA" actions serves the issue's stated motivation better than the literal AC did — the operational flow always pairs a password with a code, and the previous UI forced two clicks for one workflow. Production evidence (cited in the feature spec) supports issuing codes for unenrolled accounts, so the issue's "disabled when not enrolled" guidance was a cautious assumption that proved unnecessary.

| AC | Status | Evidence |
|---|---|---|
| `IsEnrolledIn2Sv` exposed | FULLY ADDRESSED | `WorkspaceEmailViewModels.cs`, view binding at `Accounts.cshtml:159` |
| 2FA chip per row | FULLY ADDRESSED | `Accounts.cshtml:200-213` |
| Summary counter + filter (suspended excluded) | FULLY ADDRESSED | `Accounts.cshtml:103-126`, counter logic `GoogleAdminService.cs:130-132`, filter JS `:344-361` |
| "Generate backup codes" per row, disabled-when-unenrolled | REPLACED-BY-EQUIVALENT | Replaced by Reset+2FA. Disabled-when-unenrolled gating dropped on production evidence (documented `IGoogleWorkspaceUserService.cs:54-57`, US-32.7). |
| Confirmation: "previously-issued codes will be invalidated" | NOT ADDRESSED | New confirm at `Accounts.cshtml:257` says "Both will be shown once" but omits the destructive-rotate warning. `verificationCodes.generate` is destructive — any existing paper codes are invalidated. **Minor gap, one-line fix.** |
| One-time modal display + clipboard + dismissal | FULLY ADDRESSED | `Accounts.cshtml:282-325`, TempData single-use `GoogleController.cs:762-768` |
| "Invalidate backup codes" button | NOT ADDRESSED (intentionally dropped) | Defensible — Reset+2FA rotates destructively on every use. PR body addresses obliquely. |
| Audit-logged | FULLY ADDRESSED | Both `WorkspaceAccountPasswordReset` and `WorkspaceAccountBackupCodesGenerated` written |
| Localization in `SharedResource.*.resx` | NOT ADDRESSED — rule-exempt | New copy is English-only. `memory/code/localization-admin-exempt.md` exempts admin-only views. AC literal mismatch is real but rule supersedes. |

---

### Code Quality

**IMPORTANT — Partial-success flow surfaces a green success banner saying "failed"** — `GoogleController.cs` `ResetPasswordAndGenerate2Fa` action (~line 744)

When `ResetPasswordAndGenerate2FaAsync` succeeds on the password but fails on backup-code generation, it returns `Success: true` with `BackupCode: null` and a `Message` like *"Password reset for X. Backup-code generation failed — deliver the password only..."*. The controller hits the success branch and calls `SetSuccess(result.Message)` — surfacing a **green** toast that contains the words "Backup-code generation failed". An admin may hand the user just a password without realizing the 2FA half silently didn't happen.

Fix: branch on `result.BackupCode is not null` in the controller — full-success → `SetSuccess`, partial → `SetError` (or add a `SetWarning` helper).

---

**IMPORTANT — `data-confirm` on `<button>` instead of `<form>`, breaks codebase pattern** — `Accounts.cshtml:247, 257`

All other security-sensitive forms in the codebase (`Camp/Details.cshtml`, `AdminMerge/Detail.cshtml`, `Application/Details.cshtml`) put `data-confirm` on the `<form>` element so the global `submit`-event handler catches it. The new buttons here put it on `<button>` and rely on click `preventDefault()`. Functionally works in mainstream browsers, but inconsistent on a credentials-delivery surface. Move `data-confirm` to the `<form>` on both forms.

---

**MINOR — `GenerateBackupCodesAsync` is `public` on the concrete service but not on `IGoogleAdminService`** — `GoogleAdminService.cs:325` (flagged by both reviewers)

The only intra-class caller is `ResetPasswordAndGenerate2FaAsync`. Tests reach it through the concrete type. Either declare it on the interface or make it `private`. Given no controller calls it directly, `private` is the clean answer.

---

**MINOR — Credentials live in server-rendered HTML before the modal opens** — `Accounts.cshtml:282-325`

The temp password and backup code are inline in the rendered `<pre>` and the alert-warning body. The DOM is cleared on `hidden.bs.modal`, but a print-to-PDF, browser extension, or history snapshot taken between page load and modal dismissal can capture them. The TempData + PRG pattern correctly prevents reload re-exposure but doesn't address pre-dismissal DOM exposure. Hardening only — consider rendering credentials via JS from a hidden `data-*` attribute that's cleared after first read.

---

**MINOR — `ResetPassword` action's else-branch is structurally unreachable and would emit a poor message if hit** — `GoogleController.cs:710-723`

Condition is `result.Success && !string.IsNullOrEmpty(result.TemporaryPassword)`. The "success but no password" case can't happen, but if it did, the else hits `SetError(result.ErrorMessage ?? "Failed...")` with `ErrorMessage` null. Either simplify to `if (result.Success)` (matches other controller patterns) or log defensively.

---

**MINOR — Confirmation text on Reset + 2FA missing destructive-rotate warning** — `Accounts.cshtml:257` (also a spec AC gap above)

One-line fix: append "Any previously-issued backup codes will be invalidated."

---

**Out-of-scope note (follow-up issue)** — `ResetPasswordAsync` lacks the audit-failure preservation pattern that `GenerateBackupCodesAsync` was hardened with. If audit throws after Google's password is changed, the temp password is lost and the user is locked out. Pre-existing on `main` so not a PR blocker, but the new combined flow now relies on this path.

---

### Clean Areas

- Authorization: both POST actions have `[Authorize(Policy = PolicyNames.AdminOnly)]` + `[ValidateAntiForgeryToken]`
- PRG + TempData transport for one-time credential delivery
- `Codes[0]` access guarded by `Count: > 0`
- `AuditAction.WorkspaceAccountBackupCodesInvalidated` correctly preserved as unused (stable enum values)
- New `admin.directory.user.security` scope correctly added
- Build clean, 0 warnings; all 2025 application tests pass; integration test failures are pre-existing on `main`
- `WorkspaceAccountBackupCodesGenerated` audit-after-codes pattern + empty-list guard re-verified post-redesign

---

### Bottom Line

Spec intent is fully met via a defensible reframing. Two IMPORTANT items worth fixing before merge — the green-banner-says-failed UX bug is the clearest blocker because it can mislead an admin during an active recovery. Three MINORs and one nit. No CRITICALs.

---
*Review by Claude Code `/pr-review`*

**Comment by peterdrier** [comment]
## Fixes Applied — 2026-05-04

All five in-scope items from the prior review are addressed in `0f22b1ac`. The "credentials live in DOM" hardening was intentionally skipped (judged excessive for the admin-only audience). The `ResetPasswordAsync` audit-failure-preservation note remains tracked as an out-of-scope follow-up.

**Re-review:**

- **Fix 1 — Partial-success no longer surfaces a green "failed" toast** ✓ VERIFIED
  `GoogleController.cs:744-760`. Full success (both `TempPassword` and `BackupCode`) → `SetSuccess`. Partial (password reset, backup-code failed) → `SetError` with `result.Message` ("Password reset for X, but backup-code generation failed…"). `CarryRecoveryCredentials` runs in both cases so the password modal still surfaces what was recovered. The outer `&& !string.IsNullOrEmpty(result.TempPassword)` is defensive — partial success has a non-null `TempPassword` by contract, so the partial path is reached and the inner `BackupCode` check correctly distinguishes it.

- **Fix 2 — `data-confirm` moved from `<button>` to `<form>`** ✓ VERIFIED
  Both Reset Password and Reset + 2FA forms now carry `data-confirm` on the `<form>` element, matching the established `site.js` submit-event handler pattern used in `Camp/Details.cshtml`, `AdminMerge/Detail.cshtml`, `Application/Details.cshtml`. `Accounts.cshtml:241-260`.

- **Fix 3 — `GenerateBackupCodesAsync` is now `private`** ✓ VERIFIED
  `GoogleAdminService.cs:325`. Not on `IGoogleAdminService`. The four `GenerateBackupCodesAsync_*` test cases were rerouted through `ResetPasswordAndGenerate2FaAsync` with assertions adapted to the composed result (`BackupCode` instead of `Codes`, `Message` instead of `ErrorMessage`). All 30 `GoogleAdminServiceTests` pass.

- **Fix 4 — `ResetPassword` action condition simplified** ✓ VERIFIED
  `GoogleController.cs:710` is now `if (result.Success)` with `result.TemporaryPassword!` null-forgiving (service contract guarantees non-null on success). Unreachable else-branch eliminated.

- **Fix 5 — Reset + 2FA confirmation now warns about destructive code rotation** ✓ VERIFIED
  `Accounts.cshtml:252` `data-confirm` text on the form now includes "Any previously-issued backup codes will be invalidated."

**Build/test:** Build clean, 0 warnings. All `GoogleAdminServiceTests` pass (30/30). Scope of the commit is exactly the four files in the five fix areas — no unrelated edits.

**Skipped intentionally:** Credentials-in-DOM hardening (deemed excessive for admin-only surface).

**Out of scope:** `ResetPasswordAsync` audit-failure preservation — pre-existing on `main`, not touched here.

---
*Fix & re-review by Claude Code `/pr-review`*
